File: php/list-ops/ListOps.php
<?php declare(strict_types=1);

class ListOps
{
    public function append(array $list1, array $list2): array
    {
        foreach($list2 as $item)
            $list1[] = $item;

        return $list1;
    }

    public function concat(array $list1, array ...$listn): array
    {
        foreach($listn as $list)
            foreach($list as $item)
                $list1[] = $item;

        return $list1;
    }

    /**
     * @param callable(mixed $item): bool $predicate
     */
    public function filter(callable $predicate, array $list): array
    {
        $r = [];
        foreach($list as $item) {
            if ($predicate($item))
                $r[] = $item;
        }

        return $r;
    }

    public function length(array $list): int
    {
        $len = 0;
        foreach($list as $_)
            $len++;

        return $len;
    }

    /**
     * @param callable(mixed $item): mixed $function
     */
    public function map(callable $function, array $list): array
    {
        foreach($list as $key => $item)
            $list[$key] = $function($item);

        return $list;
    }

    /**
     * @param callable(mixed $accumulator, mixed $item): mixed $function
     */
    public function foldl(callable $function, array $list, $accumulator)
    {
        foreach($list as $item)
            $accumulator = $function($accumulator, $item);

        return $accumulator;
    }

    /**
     * @param callable(mixed $accumulator, mixed $item): mixed $function
     */
    public function foldr(callable $function, array $list, $accumulator)
    {
        $rev = $this->reverse($list);
        foreach($rev as $item)
            $accumulator = $function($accumulator, $item);

        return $accumulator;
    }

    public function reverse(array $list): array
    {
        $r = [];
        foreach($list as $item)
            array_unshift($r, $item);

        return $r;
    }
}

File: python/list-ops/list_ops.py
def append(list1, list2):
    r = list1
    for i in list2:
        r.append(i)
    return r

def concat(lists):
    return [i for list in lists for i in list]

def filter(function, list):
    r = []
    for i in list:
        if function(i):
            r.append(i)
    return r

def length(list):
    a = 0
    for _ in list:
        a += 1
    return a

def map(function, list):
    r = []
    for i in list:
        r.append(function(i))
    return r

def foldl(function, list, initial):
    r = initial
    for i in list:
        r = function(r, i)
    return r


def foldr(function, list, initial):
    r = initial
    for i in reverse(list):
        r = function(r, i)
    return r

def reverse(list):
    return list[::-1]
